review/ark-blockstore-cstor/cstor.go

```diff
@@ -0,0 +1,526 @@
+	var vol *Volume
+
+	bkpname, terr := tags["ark.heptio.com/backup"]
+	if terr != true {
+		return "",  errors.New("Failed to get backup name")
+	}
+
+	if _, ret := p.volumes[volumeID]; !ret {
+		return "", errors.New("Volume is not found")
+	}
+
+	vol = p.volumes[volumeID]
+	vol.backupName = bkpname
+	err := p.backupPVC(volumeID)
+	if err != nil {
+		return "", errors.New("failed to create backup for PVC")
+	}
+
+	p.Log.Infof("creating snapshot %v", bkpname)
+
+	bkpSpec := &v1alpha1.CStorBackupSpec{
+		Name: bkpname,
+		VolumeName: volumeID,
+		CasType: casType,
+		BackupDest: p.cstorServerAddr,
+	}
+
+	bkp := &v1alpha1.CStorBackup {
+		ObjectMeta: metav1.ObjectMeta {
+			Namespace: vol.namespace,
+		},
+		Spec: *bkpSpec,
+	}
+
+        url := p.mayaAddr + backupCreatePath
+
+        bkpData, _ := json.Marshal(bkp)
+        _, err = p.httpRestCall(url, bkpData)
+        if err != nil {
+                return "", fmt.Errorf("Error calling REST api:%v", err)
+        }
+
+	p.Log.Infof("Snapshot Successfully Created")
+	filename := p.generateRemoteFilename(volumeID, vol.backupName)
+	if filename == "" {
+		return "", fmt.Errorf("Error creating remote file name for backup")
+	}
+
+	/*
+	 * For backup, we are downloading snapshot from one replica only
+	 */
+	MaxRetryCount = 1
+	ret := p.cl.UploadSnapshot(filename)
+	if ret != true {
+		return "", errors.New("Failed to upload snapshot")
+	}
+
+	return volumeID + "-ark-bkp-" + bkpname, nil
+}
+
+func (p *cstorSnapPlugin) getSnapInfo(snapshotID string) (*Snapshot, error) {
+        s := strings.Split(snapshotID, "-ark-bkp-")
+        volumeID := s[0]
+        bkpName := s[1]
+
+        pvcList, err := p.K8sClient.PersistentVolumeClaims(metav1.NamespaceAll).List(metav1.ListOptions{})
+        if err != nil {
+                return nil, fmt.Errorf("Error fetching namespaces for %s : %v", volumeID, err)
+        }
+
+        pvNs := ""
+        for _, pvc := range pvcList.Items {
+                if volumeID == pvc.Spec.VolumeName {
+                        pvNs = pvc.Namespace
+                        break
+                }
+        }
+
+        if pvNs == "" {
+                return nil, errors.New("Failed to find namespace for PVC")
+        } else {
+		return &Snapshot{
+			volID : volumeID,
+			backupName : bkpName,
+			namespace : pvNs,
+		}, nil
+	}
+}
+
+func (p *cstorSnapPlugin) CreateVolumeFromSnapshot(snapshotID, volumeType, volumeAZ string, iops *int64) (string, error) {
+	var restoreResp v1alpha1.RestoreResp
+
+	if volumeType != "cstor-snapshot" {
+		return "", fmt.Errorf("Invalid volume type(%s)", volumeType)
+	}
+
+	s := strings.Split(snapshotID, "-ark-bkp-")
+	volumeID := s[0]
+	snapName := s[1]
+
+	p.Log.Infof("Restoring snapshot %s for volume:%s", snapName, volumeID)
+
+	newVol, e := p.createPVC(volumeID, snapName)
+	if e != nil {
+		return "", fmt.Errorf("Failed to restore PVC")
+	}
+
+	p.Log.Infof("New volume(%v) created", newVol)
+
+	restoreSpec := &v1alpha1.CStorRestoreSpec{
```

Comment:
The definition of Backup & Restore schemas will be at `pkg/apis/openebs.io/backup/cstor/v1alpha1`
Corresponding CRUD operations & builder functions will be present at:
`pkg/backup/cstor/v1alpha1`

All these code will in openebs/maya repository